### PR TITLE
Fix relationship between guests, .well-known, and auth

### PR DESCRIPTION
**Pull request by turt2live**
**Reviewer**: You may be wondering what part of this giant PR needs your attention, and luckily the answer isn't as scary as the diff. The bulk of this PR has already been reviewed (see the "includes" list below), however there are two commits which do need your seal of approval (see "new commits" list below). The remainder of the PR is really just getting your approval to throw this thing on develop and see how it works in the real world, now that the entire thing has been tested by hand. This PR also serves as a changelog entry for when we release.

**Please review with https://github.com/vector-im/riot-web/pull/9779**. The js-sdk parts of this have already landed on develop.

The primary change of this PR is to require that people configure their Riot with a homeserver it can use to run the app. With that homeserver, the app is expected to use that whenever possible for guest accounts and as a default option for users. Users can change which server they ultimately use, however there's plenty of guards in place to make sure the app has a good chance of actually running once they pass the login screen.

Or written in the form of a TLDR: This is a bunch of maintenance to fix a bunch of bugs and introduce a standardized behaviour for the app's homeserver handling.

New commits (**need review**):
* https://github.com/matrix-org/matrix-react-sdk/commit/acec7f562f2f6a7316f3d4652fc3e0135939ac2c
* https://github.com/matrix-org/matrix-react-sdk/commit/7c8d322bd8672210b65fcb479353b94f54b786f3

Includes:
* https://github.com/matrix-org/matrix-react-sdk/pull/2941
* https://github.com/matrix-org/matrix-react-sdk/pull/2964
* https://github.com/matrix-org/matrix-react-sdk/pull/2965
* https://github.com/matrix-org/matrix-react-sdk/pull/2971
* https://github.com/matrix-org/matrix-react-sdk/pull/2972

Fixes:
* https://github.com/vector-im/riot-web/issues/9290 (primary issue)
* https://github.com/vector-im/riot-web/issues/8628
* https://github.com/vector-im/riot-web/issues/8825
* https://github.com/vector-im/riot-web/issues/3629
* https://github.com/vector-im/riot-web/issues/9224
* https://github.com/vector-im/riot-web/issues/9112
* https://github.com/vector-im/riot-web/issues/8826

### CODE REVIEW

**Comment by turt2live** [comment]
For completeness, this PR doesn't fix the following issues but leaves the project in a state where they could be fixed with relatively more ease:
* https://github.com/vector-im/riot-web/issues/8310
* https://github.com/vector-im/riot-web/issues/7706
* https://github.com/vector-im/riot-web/issues/9519
* https://github.com/vector-im/riot-web/issues/9084
* https://github.com/vector-im/riot-web/issues/8458
* https://github.com/vector-im/riot-web/issues/8417
* https://github.com/vector-im/riot-web/issues/8824
* https://github.com/vector-im/riot-web/issues/9563
* https://github.com/vector-im/riot-web/issues/7875
* https://github.com/matrix-org/matrix-doc/issues/1896

(most of these issues I have on my personal list of things to do when time is available, but would not object to others taking ownership of them)

File: src/components/structures/MatrixChat.js [hunk 1]
```diff
@@ -181,16 +181,8 @@ export default React.createClass({
             // Parameters used in the registration dance with the IS
```
Comment (jryans) [inline src/components/structures/MatrixChat.js]: Have you tested the email validation flow with these changes in place? (That's one area we've tended to break...)

File: src/components/views/auth/ModularServerConfig.js [hunk 2]
```diff
@@ -18,9 +18,15 @@ import React from 'react';
...
 import sdk from '../../../index';
 import { _t } from '../../../languageHandler';
+import {ValidatedServerConfig} from "../../../utils/AutoDiscoveryUtils";
+import SdkConfig from "../../../SdkConfig";
+import AutoDiscoveryUtils from "../../../utils/AutoDiscoveryUtils";
+import * as ServerType from '../../views/auth/ServerTypeSelector';
 
 const MODULAR_URL = 'https://modular.im/?utm_source=riot-web&utm_medium=web&utm_campaign=riot-web-authentication';
 
+// TODO: TravisR - Can this extend ServerConfig for most things?
```
Comment (jryans) [inline src/components/views/auth/ModularServerConfig.js]: Probably so... At the time, it seemed as if it'd be more distinct.

File: src/components/views/auth/PasswordLogin.js [hunk 3]
```diff
@@ -1,6 +1,6 @@
 /*
 Copyright 2015, 2016 OpenMarket Ltd
-Copyright 2017 Vector Creations Ltd
+Copyright 2017,2019 Vector Creations Ltd
```
Comment (jryans) [inline src/components/views/auth/PasswordLogin.js]: Seems like this should be The Matrix.org Foundation C.I.C.

File: src/utils/AutoDiscoveryUtils.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2019 New Vector Ltd
```
Comment (jryans) [inline src/utils/AutoDiscoveryUtils.js]: C.I.C.?

File: src/utils/TypeUtils.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
+/*
+Copyright 2019 New Vector Ltd
```
Comment (jryans) [inline src/utils/TypeUtils.js]: C.I.C.?

File: src/components/structures/MatrixChat.js (hunk 1 above)
Comment (turt2live) [reply src/components/structures/MatrixChat.js]: Yup, several times.

File: src/components/views/auth/PasswordLogin.js (hunk 3 above)
Comment (turt2live) [reply src/components/views/auth/PasswordLogin.js]: Changes happened the days before the asset transfer, so are still under NV

File: src/components/views/auth/ModularServerConfig.js (hunk 2 above)
Comment (turt2live) [reply src/components/views/auth/ModularServerConfig.js]: I also forgot about this comment and have stuck it on my cleanup list. That list is getting long though, so at some point I'll just take a break from feature work to fix it.

File: src/components/views/auth/PasswordLogin.js (hunk 3 above)
Comment (jryans) [reply src/components/views/auth/PasswordLogin.js]: Well, at least this one should say "New Vector Ltd" then.

Comment (turt2live) [reply src/components/views/auth/PasswordLogin.js]: Oh, yes. That's awkward.

File: src/components/structures/MatrixChat.js (hunk 1 above)
Comment (turt2live) [reply src/components/structures/MatrixChat.js]: just to follow up: I re-tested the two common scenarios just to be absolutely sure it works.

* [x] Creating a new account and verifying that email works
* [ ] Receiving an email invite does not work (see https://github.com/vector-im/riot-web/issues/9816 - this is not an issue unique to this PR)

**Comment by turt2live** [comment]
RC cut, let's land this thing 🎉

**Pull request merged**